### PR TITLE
[Java] Refactor AbstractTest & ModelResult

### PR DESCRIPTION
**Pull request by VictorGrycuk**

### CODE REVIEW

**Review by FedericoBernal — APPROVED** [review]

File: Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ModelResult.java [hunk 1]
```diff
@@ -16,5 +20,15 @@ public ModelResult(String text, int start, int end, String typeName, SortedMap<S
...
         this.resolution = resolution;
+        this.parentText = null;
+    }
+
+    public ModelResult(String text, int start, int end, String typeName, SortedMap<String, Object> resolution, @Nullable String parentText) {
+        this.text = text;
+        this.start = start;
+        this.end = end;
+        this.typeName = typeName;
+        this.resolution = resolution;
```
Comment (Batta32) [inline Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ModelResult.java]:
```suggestion
        this(text, start, end, typeName, resolution);
```

File: Java/tests/src/test/java/com/microsoft/recognizers/text/tests/AbstractTest.java [hunk 2]
```diff
@@ -265,7 +263,6 @@ public static void assertResultsWithKeys(TestCase currentCase, List<ModelResult>
         ObjectMapper mapper = new ObjectMapper();
         mapper.configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true);
         mapper.addMixIn(ModelResult.class, ModelResultMixIn.class);
-        mapper.addMixIn(ExtendedModelResult.class, ExtendedModelResultMixIn.class);
```
Comment (Batta32) [inline Java/tests/src/test/java/com/microsoft/recognizers/text/tests/AbstractTest.java:268]: What happened with the `ExtendedModelResultMixIn` class? Can you review if we need to remove it or not? The `ExtendedModelResult` was removed

File: Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ModelResult.java [hunk 3]
```diff
@@ -1,9 +1,13 @@
 package com.microsoft.recognizers.text;
 
+import com.sun.istack.internal.Nullable;
+
 import java.util.SortedMap;
 
 public class ModelResult {
 
+    public static final String ParentTextKey = "parentText";
+    public final String parentText;
```
Comment (Batta32) [inline Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ModelResult.java:10]:
```suggestion
    public final String parentText = null;
```

File: Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ModelResult.java [hunk 4]
```diff
@@ -16,5 +20,15 @@ public ModelResult(String text, int start, int end, String typeName, SortedMap<S
         this.end = end;
         this.typeName = typeName;
         this.resolution = resolution;
+        this.parentText = null;
```
Comment (Batta32) [inline Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ModelResult.java]:
```suggestion
```

**Review by Batta32 — CHANGES_REQUESTED** [review]
I'm not quite sure to review the `ExtendedModelResult` as this will be a disparity against C#/JS/Python (see [this query](https://github.com/microsoft/Recognizers-Text/search?q=parentText)).

File: Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ModelResult.java [hunk 5]
```diff
@@ -1,20 +1,29 @@
 package com.microsoft.recognizers.text;
 
+import com.sun.istack.internal.Nullable;
+
```
Comment (Batta32) [inline Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ModelResult.java]: Can we remove this? We are not using this

File: Java/tests/src/test/java/com/microsoft/recognizers/text/tests/AbstractTest.java [hunk 6]
```diff
@@ -158,17 +152,26 @@ public static void assertResultsWithKeys(TestCase currentCase, List<ModelResult>
...
+    protected void assertModel(ModelResult expected, ModelResult actual, TestCase currentCase, List<String> testResolutionKeys) {
+        if (expected.resolution.containsKey(ResolutionKey.Value)) {
+            Assert.assertEquals(getMessage(currentCase, "resolution.value"),
+                    expected.resolution.get(ResolutionKey.Value), actual.resolution.get(ResolutionKey.Value));
+        }
+
+        for (String key : testResolutionKeys) {
+            Assert.assertEquals(getMessage(currentCase, key), expected.resolution.get(key), actual.resolution.get(key));
+        }
+    }
```
Comment (Batta32) [inline Java/tests/src/test/java/com/microsoft/recognizers/text/tests/AbstractTest.java]: Move this method to after the `public` ones (public -> protected -> private)

File: Java/tests/src/test/java/com/microsoft/recognizers/text/tests/AbstractTest.java (hunk 2 above)
Comment (Batta32) [reply Java/tests/src/test/java/com/microsoft/recognizers/text/tests/AbstractTest.java:268]: Please, let me know about this, I want to know if we need to remove the class, or if we need to apply another change for this removal

**Review by Batta32 — CHANGES_REQUESTED** [review]

File: Java/tests/src/test/java/com/microsoft/recognizers/text/tests/AbstractTest.java [hunk 7]
```diff
@@ -152,6 +152,7 @@ public void assertResults(TestCase currentCase, List<ModelResult> results, List<
                     Assert.assertEquals(getMessage(currentCase, "typeName"), expected.typeName, actual.typeName);
                     Assert.assertEquals(getMessage(currentCase, "text"), expected.text, actual.text);
 
+                    // If both properties are equal to 0, it means they doesn't exists, so it will skip the validation
```
Comment (VictorGrycuk) [inline Java/tests/src/test/java/com/microsoft/recognizers/text/tests/AbstractTest.java]: `they doesn't` -> `they don't`

Comment (VictorGrycuk) [reply Java/tests/src/test/java/com/microsoft/recognizers/text/tests/AbstractTest.java]: Also, `exists` -> `exist`

**Review by Batta32 — CHANGES_REQUESTED** [review]
I agree to remove the ExtendedModelResult and use the ModelResult, and also to remove the ExtendedModelResultMixIn, but we need to change the reference of [this](https://github.com/southworks/Recognizers-Text/pull/376/files#diff-2c7744f4662ec42f5af8566f29f0392686661d22370ad2bf1dc5ac64eb6ed5dfL268), to use the [ModelResultMixIn](https://github.com/microsoft/Recognizers-Text/blob/08b488ae1c83127eb93b26d381b8d668b548a620/Java/tests/src/test/java/com/microsoft/recognizers/text/tests/helpers/ModelResultMixIn.java) adding the `@JsonProperty("parentText") String parentText;`

It's weird to remove that mapper.

File: Java/tests/src/test/java/com/microsoft/recognizers/text/tests/AbstractTest.java (hunk 2 above)
Comment (Batta32) [reply Java/tests/src/test/java/com/microsoft/recognizers/text/tests/AbstractTest.java:268]: Applying this [change](https://github.com/southworks/Recognizers-Text/pull/376/files#diff-b3c211fd72ec7940e1654281c36961948b46dc6445a0485f7a2da87a08924332R13) requires to update this as follows:
`mapper.addMixIn(ModelResult.class, ModelResultMixIn.class);`

Also, let us know if the unit tests are still working after these changes

**Review by Batta32 — CHANGES_REQUESTED** [review]

**Review by Batta32 — APPROVED** [review]

File: Java/tests/src/test/java/com/microsoft/recognizers/text/tests/datetime/DateTimeTest.java [hunk 8]
```diff
@@ -35,45 +35,29 @@ public DateTimeTest(TestCase currentCase) {
...
-                    if (actual instanceof ExtendedModelResult) {
-                        Assert.assertEquals(getMessage(currentCase, "parentText"),
-                                            expected.parentText, ((ExtendedModelResult)actual).parentText);
-                    }
+    protected void assertModel(ModelResult expected,
+                               ModelResult actual,
+                               TestCase currentCase,
+                               List<String> testResolutionKeys) {
+        Assert.assertEquals(getMessage(currentCase, "start"), expected.start, actual.start);
+        Assert.assertEquals(getMessage(currentCase, "end"), expected.end, actual.end);
```
Comment (Batta32) [inline Java/tests/src/test/java/com/microsoft/recognizers/text/tests/datetime/DateTimeTest.java]: This is already validated in `AbstractTest` right? Can we safely remove it?

**Review by Batta32 — CHANGES_REQUESTED** [review]

File: Java/tests/src/test/java/com/microsoft/recognizers/text/tests/datetime/DateTimeTest.java (hunk 8 above)
Comment (Batta32) [reply Java/tests/src/test/java/com/microsoft/recognizers/text/tests/datetime/DateTimeTest.java]: Removed, they are validated in `AbstractTest` class

**Pull request closed without merging**